README.md

```diff
@@ -1,277 +1,899 @@
+The offer beacon asset name is `sha2_256( offer_policy_id ++ offer_asset_name )`.
 
-The Cardano-Swaps Protocol is broadly comprised of two steps:
+The trading pair beacon asset name is 
+`sha2_256( offer_id ++ offer_name ++ ask_id ++ ask_name )`. However, in the case where one
+of the assets is ADA (which is just the empty bytestring), the ADA policy id is set to "00" instead.
+Without this changed, there would be no difference between the offer beacon and trading pair beacon
+when ADA is part of a swap.
 
-1. **Prepare Swap Address** - Alice prepares a swap address for swapping Token "ABC" for Token "XYZ" by minting an ABC:XYZ Beacon Token in a "Beacon UTxO" that designates her address specifically as an ABC-for-XYZ swap address. She then (in the same or separate transaction) outputs one or more "Swap UTxO(s)" with "ABC" Tokens, with an attached `Price` Datum that designates her desired price (in "XYZ") for that particular UTxO.
-   
-2. **Execute Swap(s)** - Bob queries a list of ABC-for-XYZ swap addresses (via the ABC:XYZ Beacon Token), and finds some attractively priced UTxOs in Alice's swap address. He submits a transaction that consumes one (or multiple) of Alice's "Swap UTxO(s)", and one or more of his own UTxO(s) containing Alice's requested "XYZ" Token. Bob can then output Alice's "ABC" Tokens to himself in proportion to the amount of "XYZ" tokens he outputs back to Alice. Bob does not have to buy all of the contents of one of Alice's UTxOs - he buy however much he wants, and output the remaining "ABC" back to Alice in the same UTxO as her "XYZ" payment. This UTxO (containing Alice's "XYZ" tokens) remains swappable as long as some "ABC" tokens remain in it; nobody except Alice can claim the "XYZ" tokens.
+These beacons allow for expressive queries since it makes it possible to query all swaps where XYZ
+is being offered, regardless of the asset being asked for. This is in addition to being able to
+query all swaps for a specific trading pair. This extra expressiveness improves the arbitrager's
+ability to find profitable paths through currently open swaps.
 
-These features are further explained in the [Discussion section](<#Discussion & FAQ>) below
+##### Swap Address
 
-**This Aiken version of Cardano-Swaps can handle up to 14 swaps in a single transaction with a fee of only 1.5 ADA. A transaction with a single swap only costs 0.25 ADA. This makes Cardano-Swaps up to an order of magnitude cheaper than existing Cardano DEXs.** See [benchmarks](Benchmarks.md) for details.
+All users get a single swap address where all of their one-way swaps will be located. This universal
+validator can enforce all possible swaps for the user. And since users only need to worry about a
+single address, it keeps the UI simple.
```

Comment:
> One of the reasons Cardano wallets choose new addresses is to provide quantum resistance because only the hash of the public key is known until the first eUTxO is spent from that address.

I did not know this. Can choosing different staking credentials be enough to provide the same resistance? You can always get a different address if you use a different staking credential.

Reply:
No, as I understand different staking credentials won't help with quantum resistance because you will have already executed a transaction that required exposing your public key when you stake with a pool.

I could have this a bit wrong, but here is how I understand this quantum resistance feature to work.  Normal addresses on Cardano are actually the hash of the public key.  Since the hash used is quantum resistant, this means that the public key is not actually known until a transaction is actually executed for that address.  When a transaction is executed, you provide a copy of your public key + signature using private key and these are checked firstly to confirm that the signature is correct and then also that the public key hash which represents the address is correct.

This is why wallets like Yoroi, daedalus and others generate a new address each time when you click on the receive funds tab.  The designers were tricky in using the hash of the public key as the address rather than just the public key itself.  The hash used is apparently quantum resistant, but ed25519 keys are supposedly not.

Reply:
Yes, the quantum resistance effect by creating new addresses which are the hash of an unused public key is a nice trick but it is very clear that we will need updated cryptology when quantum computers get closer.